### PR TITLE
Prerequisites for highlighter (1)

### PR DESCRIPTION
**Pull request by mnunberg**
This contains various API changes needed for highlighting to work properly.

### CODE REVIEW

File: src/forward_index.c [hunk 1]
```diff
@@ -142,17 +142,14 @@ static khIdxEntry *makeEntry(ForwardIndex *idx, const char *s, size_t n, uint32_
   return (khIdxEntry *)bb;
 }
 
-// void ForwardIndex_NormalizeFreq(ForwardIndex *idx, ForwardIndexEntry *e) {
-//   e->freq = e->freq / idx->maxFreq;
-// }
-int forwardIndexTokenFunc(void *ctx, const Token *t) {
-  ForwardIndex *idx = ctx;
-
+static void ForwardIndex_HandleToken(ForwardIndex *idx, const char *tok, size_t tokLen,
```
Comment (dvirsky) [inline src/forward_index.c]: why drop the token?

File: src/tests/test_index.c [hunk 2]
```diff
@@ -615,12 +615,10 @@ typedef struct {
 
 int tokenFunc(void *ctx, const Token *t) {
   tokenContext *tx = ctx;
-  int ret = strcmp(t->s, tx->expected[tx->num++]);
+  int ret = strncmp(t->tok, tx->expected[tx->num++], t->tokLen);
+  printf("Compare: %.*s <=> %s\n", (int)t->tokLen, t->tok, tx->expected[tx->num - 1]);
```
Comment (dvirsky) [inline src/tests/test_index.c]: remove this obviously

File: src/toksep.h [hunk 3]
```diff
@@ -0,0 +1,38 @@
+#ifndef TOKSEP_H
+#define TOKSEP_H
+
+#include <stdint.h>
+#include <stdlib.h>
+//! " # $ % & ' ( ) * + , - . / : ; < = > ? @ [ \ ] ^ _ ` { | } ~
```
Comment (dvirsky) [inline src/toksep.h]: underscore is no longer a separator though. we can remove it from the comment

File: src/forward_index.c (hunk 1 above)
Comment (mnunberg) [reply src/forward_index.c]: Micro optimization to avoid having to re-read the struct field again.

**Comment by mnunberg** [comment]
Updated description:

This PR contains mostly 'static' or seemingly cosmetic changes required for the highlighter systems to work. These changes should make the API better and faster overall.

The main components changed are:

### Tokenizer

* The tokenizer should have a mode that does not modify the document being tokenized. When performing fragmenting, we need to have access to the original document, and don't want to make a copy.
* I've removed some unused fields and dead code.
* The structure passed to the callback now also includes a pointer to the raw text in the source document. This is used to record the byte offset for the given term.
* Tokenizing is now passed a context object containing other information, such as pointer to doc, etc. and basically other information that is not stateful to the tokenizer itself.

### Query Serialization/Parsing

* Removed some legacy functions superseded by RMUtil
* Provided an abstraction for returned fields. This contains various constants related to highlighting, but these aren't used in the PR yet.
* Split serialization from search. Serialization seems small, but in the future commits, serialization will get larger as we include highlighting stuff.

### Array

I've added an 'Array' data type which is basically a thin wrapper over a C array. This is designed for C structs and is lower level than the Buffer, etc. API.

**Comment by mnunberg** [comment]
squashed

**Pull request closed without merging**